src/SparseEnsembleKalmanInversion.jl

```diff
@@ -0,0 +1,153 @@
+    h = fill(FT(0), 2 * N_params, 1)
+    G1 = vcat(G, hcat(fill(FT(0), 1, size(P)[1]), fill(FT(1), 1, N_params)))
+    h1 = vcat(h, γ)
+    x = Variable(size(P1)[1])
+    problem = minimize(0.5 * quadform(x, P1; assume_psd = true) + q1' * x, [G1 * x <= h1])
+    solve!(problem, () -> SCS.Optimizer(verbose = false))
+
+    return hcat(H_u, fill(FT(0), size(H_u)[1], N_params)) * evaluate(x)
+end
+
+"""
+    update_ensemble!(ekp::EnsembleKalmanProcess{FT, IT, <:SparseInversion{FT,IT}}, g::Array{FT,2} cov_threshold::FT=0.01, Δt_new=nothing) where {FT, IT}
+
+Updates the ensemble according to which type of Process we have. Model outputs `g` need to be a `N_obs × N_ens` array (i.e data are columms).
+"""
+function update_ensemble!(
+    ekp::EnsembleKalmanProcess{FT, IT, SparseInversion{FT, IT}},
+    g::Array{FT, 2};
+    cov_threshold::FT = 0.01,
+    Δt_new = nothing,
+    deterministic_forward_map = true,
+) where {FT, IT}
+
+    # Update follows eqns. (4) and (5) of Schillings and Stuart (2017)
+
+    #catch works when g non-square
+    if !(size(g)[2] == ekp.N_ens)
+        throw(DimensionMismatch("ensemble size in EnsembleKalmanProcess and g do not match, try transposing g or check ensemble size"))
+    end
+
+    # u: N_par × N_ens 
+    # g: N_obs × N_ens
+    u = get_u_final(ekp)
+    N_obs = size(g, 1)
+
+    cov_init = cov(u, dims = 2)
+    cov_ug = cov(u, g, dims = 2, corrected = false) # [N_par × N_obs]
+    cov_gg = cov(g, g, dims = 2, corrected = false) # [N_obs × N_obs]
+    cov_uu = cov(u, u, dims = 2, corrected = false) # [N_par × N_par]
+
+    if !isnothing(Δt_new)
+        push!(ekp.Δt, Δt_new)
+    elseif isnothing(Δt_new) && isempty(ekp.Δt)
+        push!(ekp.Δt, FT(1))
+    else
+        push!(ekp.Δt, ekp.Δt[end])
+    end
+
+    v = hcat(u', g')
+
+    # Scale noise using Δt
+    scaled_obs_noise_cov = ekp.obs_noise_cov / ekp.Δt[end]
+    noise = rand(MvNormal(zeros(N_obs), scaled_obs_noise_cov), ekp.N_ens)
+
+    # Add obs_mean (N_obs) to each column of noise (N_obs × N_ens) if
+    # G is deterministic
+    y = deterministic_forward_map ? (ekp.obs_mean .+ noise) : (ekp.obs_mean .+ zero(noise))
+
+    # N_obs × N_obs \ [N_obs × N_ens]
+    # --> tmp is [N_obs × N_ens]
+    tmp = (cov_gg + scaled_obs_noise_cov) \ (y - g)
+    u += (cov_ug * tmp) # [N_par × N_ens]
+
+    # store new parameters (and model outputs)
+    push!(ekp.u, DataContainer(u, data_are_columns = true))
+    push!(ekp.g, DataContainer(g, data_are_columns = true))
+
+    # Sparse EKI
+    cov_vv = vcat(hcat(cov_uu, cov_ug), hcat(cov_ug', cov_gg))
+    H_u = hcat(1.0 * I(size(u)[1]), fill(FT(0), size(u)[1], size(g)[1]))
+    H_g = hcat(fill(FT(0), size(g)[1], size(u)[1]), 1.0 * I(size(g)[1]))
+
+    H_uc = H_u[ekp.process.uc_idx, :]
+
+    Threads.@threads for j in 1:(ekp.N_ens)
+        # Solve a quadratic programming problem
+        u[:, j] = sparse_qp(ekp, v[j, :], inv(cov_vv), H_u, H_g, y[:, j], ekp.process.γ, H_uc = H_uc)
```

Comment:
Can you pass `cov_vv` instead of `inv(cov_vv)` and solve the linear system in `sparse_qp`?